Source/Core/DolphinQt2/Settings/USBDeviceAddToWhitelistDialog.cpp

```diff
@@ -0,0 +1,169 @@
+{
+  InitControls();
+  setLayout(main_layout);
+}
+
+void USBDeviceAddToWhitelistDialog::InitControls()
+{
+  setWindowTitle(tr("Add New USB Device"));
+  whitelist_buttonbox = new QDialogButtonBox();
+  auto* add_button = new QPushButton(tr("Add..."));
+  whitelist_buttonbox->addButton(add_button, QDialogButtonBox::NoRole);
+  connect(add_button, &QPushButton::released, this,
+          &USBDeviceAddToWhitelistDialog::AddUSBDeviceToWhitelist);
+  add_button->setDefault(true);
+
+  main_layout = new QVBoxLayout();
+  enter_device_id_label = new QLabel(tr("Enter USB Device ID"));
+  enter_device_id_label->setAlignment(Qt::AlignCenter);
+  main_layout->addWidget(enter_device_id_label);
+
+  entry_form_layout = new QFormLayout();
+  entry_form_layout->setFieldGrowthPolicy(QFormLayout::FieldsStayAtSizeHint);
+  entry_form_layout->setLabelAlignment(Qt::AlignJustify | Qt::AlignVCenter);
+  entry_form_layout->setFormAlignment(Qt::AlignHCenter | Qt::AlignTop);
+  device_vid_textbox = new QLineEdit();
+  QSizePolicy sizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::Fixed);
+  sizePolicy.setHorizontalStretch(1);
+  sizePolicy.setVerticalStretch(0);
+  sizePolicy.setHeightForWidth(device_vid_textbox->sizePolicy().hasHeightForWidth());
+  device_vid_textbox->setSizePolicy(sizePolicy);
+
+  entry_form_layout->setWidget(2, QFormLayout::LabelRole, device_vid_textbox);
+
+  device_pid_textbox = new QLineEdit();
+  sizePolicy.setHeightForWidth(device_pid_textbox->sizePolicy().hasHeightForWidth());
+  device_pid_textbox->setSizePolicy(sizePolicy);
+
+  entry_form_layout->setWidget(2, QFormLayout::FieldRole, device_pid_textbox);
+  main_layout->addLayout(entry_form_layout);
+
+  select_label = new QLabel(tr("or select a device"));
+  select_label->setAlignment(Qt::AlignCenter);
+
+  main_layout->addWidget(select_label);
+
+  usb_inserted_devices_list = new QListWidget();
+  m_refresh_devices_timer = new QTimer(this);
+  connect(usb_inserted_devices_list, &QListWidget::currentItemChanged, this,
+          &USBDeviceAddToWhitelistDialog::OnDeviceSelection);
+  connect(m_refresh_devices_timer, &QTimer::timeout, this,
+          &USBDeviceAddToWhitelistDialog::OnRefreshDevicesTimer);
+  m_refresh_devices_timer->start(1000);
+
+  main_layout->addWidget(usb_inserted_devices_list);
+  main_layout->addWidget(whitelist_buttonbox);
+
+  // i18n: VID means Vendor ID (in the context of a USB device)
+  device_vid_textbox->setPlaceholderText(tr("Device VID (e.g., 057e)"));
+  // i18n: PID means Product ID (in the context of a USB device), not Process ID
+  device_pid_textbox->setPlaceholderText(tr("Device PID (e.g., 0305)"));
+}
+
+void USBDeviceAddToWhitelistDialog::RefreshDeviceList()
+{
+  const auto& current_devices = USBUtils::GetInsertedDevices();
+  if (current_devices == m_shown_devices)
+    return;
+  QSignalBlocker blocker(usb_inserted_devices_list);  // unblocks when out of scope.
```

Comment:
Blocking signals seems heavy-handed.

Reply:
You're right, I don't think it is needed at all.